server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -801,6 +806,18 @@ private PipelineIterator getAndResetPipelines(IndexRequest indexRequest) {
         return new PipelineIterator(pipelineId, finalPipelineId);
     }
 
+    /**
+     * Returns the first pipeline to be run in the request; does not modify the request
+     */
+    private Pipeline getFirstPipeline(IndexRequest indexRequest) {
+        if (indexRequest.getPipeline().equals(NOOP_PIPELINE_NAME) == false) {
```

Comment:
I think you're right, but I'm wondering if it would it be sufficient to change these lines to `NOOP_PIPELINE_NAME.equals(indexRequest.getPipeline())` etc? The `PipelineIterator` works more like that, as does the `hasPipeline(...)` method.

Reply:
I've been playing around with this code some and see that `indexRequest.getPipeline()` and `getFinalPipeline()` are never actually null, so this code already works. But, I think it would be more obvious that there isn't a chance of an NPE if we flip the equality check to `NOOP_PIPELINE_NAME.equals(indexRequest.getPipeline())` as you described; which will work well. 

Alternatively, we could move this logic into `PipelineIterator`, and make use of the `NOOP...` check that it already does. Eg, we could add something like this to `PipelineIterator`
```
 public Pipeline peekFirst() {
            return getPipeline(defaultPipeline != null ? defaultPipeline : finalPipeline);
 }
```
then set `firstPipeline` a few lines down after creating `PipelineIterator`. 

I am fine with this, or the existing method if you think it doesn't belong in `PipelineIterator`

Reply:
I think that makes a lot of sense, I'll put it in `PipelineIterator`